In [31]:
import pandas as pd
import tweepy
import requests
import json
import os
import zipfile
import time as timer
import datetime

In [6]:
twitter_posts = pd.read_csv(r"C:\Users\HP\Downloads\twitter-archive-enhanced.csv", sep=',')

In [7]:
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)
with open(os.path.join(url.split('/')[-1]), mode = 'wb') as file:
    file.write(response.content)
    
img_predict = pd.read_csv('image-predictions.tsv', sep='\t')

In [8]:
os.listdir()

['.ipynb_checkpoints',
 'Data Wrangling - Project 2(Udacity).ipynb',
 'image-predictions.tsv']

In [9]:
bearer_token = "AAAAAAAAAAAAAAAAAAAAAAvjeAEAAAAA%2BFW5RQwLXvUwBECFlR3lGyq%2Btc0%3DyaeSH8y76detA8avly1rSUmnRliHfryEDwNuxR9hpbxGEJrWpe"

api_key = "JKyFq5aiWuYOqUMTvBfc7OkWT"

api_key_secret = "RIA9FaUfB81UtHdqUXqsYjZKVaJWrtnKIzVmYqGkS7rr5JaIOR"

access_token = "215757439-VaMiGSXBLeSL9ZW3uyQyIKleC80Ifqh4XKTTCgGA"

access_token_secret = "B17HUwjAB6r8zmIBHVHSOiSmevasmoNnGgcI6v8KN4wwC"

auth = tweepy.OAuthHandler(api_key, api_key_secret)

auth.set_access_token(access_token, access_token_secret)

api = tweepy.API(auth, wait_on_rate_limit = True)

try:
    api.verify_credentials()
    print('Authentication is Successful')
except:
    print('Authentication Failed')

Authentication is Successful


In [10]:
tweet_ids = twitter_posts.tweet_id.values
len(tweet_ids)
count = 0
failed_id = {}
with open('tweet_json.txt', 'w') as outfile:
    for tweet_id in tweet_ids:
        count += 1
        print(str(count)+ ":" + str(tweet_id))
        try:
            tweet = api.get_status(tweet_id, tweet_mode = 'extended')
            print('Succeeded in retrieving this tweet')
            json.dump(tweet._json, outfile)
            outfile.write('\n')
        except tweepy.TweepyException as e:
            print('Failed to retrieve tweets with id')
            failed_id[tweet_id] = e
            pass
print(end - start)
print(failed_id)

1:892420643555336193
Succeeded in retrieving this tweet
2:892177421306343426
Succeeded in retrieving this tweet
3:891815181378084864
Succeeded in retrieving this tweet
4:891689557279858688
Succeeded in retrieving this tweet
5:891327558926688256
Succeeded in retrieving this tweet
6:891087950875897856
Succeeded in retrieving this tweet
7:890971913173991426
Succeeded in retrieving this tweet
8:890729181411237888
Succeeded in retrieving this tweet
9:890609185150312448
Succeeded in retrieving this tweet
10:890240255349198849
Succeeded in retrieving this tweet
11:890006608113172480
Succeeded in retrieving this tweet
12:889880896479866881
Succeeded in retrieving this tweet
13:889665388333682689
Succeeded in retrieving this tweet
14:889638837579907072
Succeeded in retrieving this tweet
15:889531135344209921
Succeeded in retrieving this tweet
16:889278841981685760
Succeeded in retrieving this tweet
17:888917238123831296
Succeeded in retrieving this tweet
18:888804989199671297
Succeeded in retri

Succeeded in retrieving this tweet
145:863907417377173506
Succeeded in retrieving this tweet
146:863553081350529029
Succeeded in retrieving this tweet
147:863471782782697472
Succeeded in retrieving this tweet
148:863432100342583297
Succeeded in retrieving this tweet
149:863427515083354112
Succeeded in retrieving this tweet
150:863079547188785154
Succeeded in retrieving this tweet
151:863062471531167744
Succeeded in retrieving this tweet
152:862831371563274240
Succeeded in retrieving this tweet
153:862722525377298433
Succeeded in retrieving this tweet
154:862457590147678208
Succeeded in retrieving this tweet
155:862096992088072192
Succeeded in retrieving this tweet
156:861769973181624320
Failed to retrieve tweets with id
157:861383897657036800
Succeeded in retrieving this tweet
158:861288531465048066
Succeeded in retrieving this tweet
159:861005113778896900
Succeeded in retrieving this tweet
160:860981674716409858
Succeeded in retrieving this tweet
161:860924035999428608
Succeeded in re

Succeeded in retrieving this tweet
287:838831947270979586
Succeeded in retrieving this tweet
288:838561493054533637
Succeeded in retrieving this tweet
289:838476387338051585
Succeeded in retrieving this tweet
290:838201503651401729
Succeeded in retrieving this tweet
291:838150277551247360
Succeeded in retrieving this tweet
292:838085839343206401
Succeeded in retrieving this tweet
293:838083903487373313
Succeeded in retrieving this tweet
294:837820167694528512
Succeeded in retrieving this tweet
295:837482249356513284
Succeeded in retrieving this tweet
296:837471256429613056
Succeeded in retrieving this tweet
297:837366284874571778
Failed to retrieve tweets with id
298:837110210464448512
Succeeded in retrieving this tweet
299:837012587749474308
Failed to retrieve tweets with id
300:836989968035819520
Succeeded in retrieving this tweet
301:836753516572119041
Succeeded in retrieving this tweet
302:836677758902222849
Succeeded in retrieving this tweet
303:836648853927522308
Succeeded in ret

Succeeded in retrieving this tweet
429:821149554670182400
Succeeded in retrieving this tweet
430:821107785811234820
Succeeded in retrieving this tweet
431:821044531881721856
Succeeded in retrieving this tweet
432:820837357901512704
Succeeded in retrieving this tweet
433:820749716845686786
Succeeded in retrieving this tweet
434:820690176645140481
Succeeded in retrieving this tweet
435:820494788566847489
Succeeded in retrieving this tweet
436:820446719150292993
Succeeded in retrieving this tweet
437:820314633777061888
Succeeded in retrieving this tweet
438:820078625395449857
Succeeded in retrieving this tweet
439:820013781606658049
Succeeded in retrieving this tweet
440:819952236453363712
Succeeded in retrieving this tweet
441:819924195358416896
Succeeded in retrieving this tweet
442:819711362133872643
Succeeded in retrieving this tweet
443:819588359383371776
Succeeded in retrieving this tweet
444:819347104292290561
Succeeded in retrieving this tweet
445:819238181065359361
Succeeded in r

Succeeded in retrieving this tweet
571:801854953262350336
Succeeded in retrieving this tweet
572:801538201127157760
Succeeded in retrieving this tweet
573:801285448605831168
Succeeded in retrieving this tweet
574:801167903437357056
Succeeded in retrieving this tweet
575:801127390143516673
Succeeded in retrieving this tweet
576:801115127852503040
Succeeded in retrieving this tweet
577:800859414831898624
Succeeded in retrieving this tweet
578:800855607700029440
Succeeded in retrieving this tweet
579:800751577355128832
Succeeded in retrieving this tweet
580:800513324630806528
Succeeded in retrieving this tweet
581:800459316964663297
Succeeded in retrieving this tweet
582:800443802682937345
Succeeded in retrieving this tweet
583:800388270626521089
Succeeded in retrieving this tweet
584:800188575492947969
Succeeded in retrieving this tweet
585:800141422401830912
Succeeded in retrieving this tweet
586:800018252395122689
Succeeded in retrieving this tweet
587:799774291445383169
Succeeded in r

Succeeded in retrieving this tweet
713:784431430411685888
Succeeded in retrieving this tweet
714:784183165795655680
Succeeded in retrieving this tweet
715:784057939640352768
Succeeded in retrieving this tweet
716:783839966405230592
Succeeded in retrieving this tweet
717:783821107061198850
Succeeded in retrieving this tweet
718:783695101801398276
Succeeded in retrieving this tweet
719:783466772167098368
Succeeded in retrieving this tweet
720:783391753726550016
Succeeded in retrieving this tweet
721:783347506784731136
Succeeded in retrieving this tweet
722:783334639985389568
Succeeded in retrieving this tweet
723:783085703974514689
Succeeded in retrieving this tweet
724:782969140009107456
Succeeded in retrieving this tweet
725:782747134529531904
Succeeded in retrieving this tweet
726:782722598790725632
Succeeded in retrieving this tweet
727:782598640137187329
Succeeded in retrieving this tweet
728:782305867769217024
Succeeded in retrieving this tweet
729:782021823840026624
Succeeded in r

Succeeded in retrieving this tweet
855:765222098633691136
Succeeded in retrieving this tweet
856:764857477905154048
Succeeded in retrieving this tweet
857:764259802650378240
Succeeded in retrieving this tweet
858:763956972077010945
Succeeded in retrieving this tweet
859:763837565564780549
Succeeded in retrieving this tweet
860:763183847194451968
Succeeded in retrieving this tweet
861:763167063695355904
Succeeded in retrieving this tweet
862:763103485927849985
Succeeded in retrieving this tweet
863:762699858130116608
Succeeded in retrieving this tweet
864:762471784394268675
Succeeded in retrieving this tweet
865:762464539388485633
Succeeded in retrieving this tweet
866:762316489655476224
Succeeded in retrieving this tweet
867:762035686371364864
Succeeded in retrieving this tweet
868:761976711479193600
Succeeded in retrieving this tweet
869:761750502866649088
Succeeded in retrieving this tweet
870:761745352076779520
Succeeded in retrieving this tweet
871:761672994376806400
Succeeded in r

Rate limit reached. Sleeping for: 272


Succeeded in retrieving this tweet
902:758474966123810816
Succeeded in retrieving this tweet
903:758467244762497024
Succeeded in retrieving this tweet
904:758405701903519748
Succeeded in retrieving this tweet
905:758355060040593408
Succeeded in retrieving this tweet
906:758099635764359168
Succeeded in retrieving this tweet
907:758041019896193024
Succeeded in retrieving this tweet
908:757741869644341248
Succeeded in retrieving this tweet
909:757729163776290825
Succeeded in retrieving this tweet
910:757725642876129280
Succeeded in retrieving this tweet
911:757611664640446465
Succeeded in retrieving this tweet
912:757597904299253760
Succeeded in retrieving this tweet
913:757596066325864448
Succeeded in retrieving this tweet
914:757400162377592832
Succeeded in retrieving this tweet
915:757393109802180609
Succeeded in retrieving this tweet
916:757354760399941633
Succeeded in retrieving this tweet
917:756998049151549440
Succeeded in retrieving this tweet
918:756939218950160384
Succeeded in r

Succeeded in retrieving this tweet
1043:743895849529389061
Succeeded in retrieving this tweet
1044:743835915802583040
Succeeded in retrieving this tweet
1045:743609206067040256
Succeeded in retrieving this tweet
1046:743595368194129920
Succeeded in retrieving this tweet
1047:743545585370791937
Succeeded in retrieving this tweet
1048:743510151680958465
Succeeded in retrieving this tweet
1049:743253157753532416
Succeeded in retrieving this tweet
1050:743222593470234624
Succeeded in retrieving this tweet
1051:743210557239623680
Succeeded in retrieving this tweet
1052:742534281772302336
Succeeded in retrieving this tweet
1053:742528092657332225
Succeeded in retrieving this tweet
1054:742465774154047488
Succeeded in retrieving this tweet
1055:742423170473463808
Succeeded in retrieving this tweet
1056:742385895052087300
Succeeded in retrieving this tweet
1057:742161199639494656
Succeeded in retrieving this tweet
1058:742150209887731712
Succeeded in retrieving this tweet
1059:7417932638128087

Succeeded in retrieving this tweet
1182:719332531645071360
Succeeded in retrieving this tweet
1183:718971898235854848
Succeeded in retrieving this tweet
1184:718939241951195136
Succeeded in retrieving this tweet
1185:718631497683582976
Succeeded in retrieving this tweet
1186:718613305783398402
Succeeded in retrieving this tweet
1187:718540630683709445
Succeeded in retrieving this tweet
1188:718460005985447936
Succeeded in retrieving this tweet
1189:718454725339934721
Succeeded in retrieving this tweet
1190:718246886998687744
Succeeded in retrieving this tweet
1191:718234618122661888
Succeeded in retrieving this tweet
1192:717841801130979328
Succeeded in retrieving this tweet
1193:717790033953034240
Succeeded in retrieving this tweet
1194:717537687239008257
Succeeded in retrieving this tweet
1195:717428917016076293
Succeeded in retrieving this tweet
1196:717421804990701568
Succeeded in retrieving this tweet
1197:717047459982213120
Succeeded in retrieving this tweet
1198:7170093624520908

Succeeded in retrieving this tweet
1321:706346369204748288
Succeeded in retrieving this tweet
1322:706310011488698368
Succeeded in retrieving this tweet
1323:706291001778950144
Succeeded in retrieving this tweet
1324:706265994973601792
Succeeded in retrieving this tweet
1325:706169069255446529
Succeeded in retrieving this tweet
1326:706166467411222528
Succeeded in retrieving this tweet
1327:706153300320784384
Succeeded in retrieving this tweet
1328:705975130514706432
Succeeded in retrieving this tweet
1329:705970349788291072
Succeeded in retrieving this tweet
1330:705898680587526145
Succeeded in retrieving this tweet
1331:705786532653883392
Succeeded in retrieving this tweet
1332:705591895322394625
Succeeded in retrieving this tweet
1333:705475953783398401
Succeeded in retrieving this tweet
1334:705442520700944385
Succeeded in retrieving this tweet
1335:705428427625635840
Succeeded in retrieving this tweet
1336:705239209544720384
Succeeded in retrieving this tweet
1337:7052234446868889

Succeeded in retrieving this tweet
1460:695064344191721472
Succeeded in retrieving this tweet
1461:695051054296211456
Succeeded in retrieving this tweet
1462:694925794720792577
Succeeded in retrieving this tweet
1463:694905863685980160
Succeeded in retrieving this tweet
1464:694669722378485760
Succeeded in retrieving this tweet
1465:694356675654983680
Succeeded in retrieving this tweet
1466:694352839993344000
Succeeded in retrieving this tweet
1467:694342028726001664
Succeeded in retrieving this tweet
1468:694329668942569472
Succeeded in retrieving this tweet
1469:694206574471057408
Succeeded in retrieving this tweet
1470:694183373896572928
Succeeded in retrieving this tweet
1471:694001791655137281
Succeeded in retrieving this tweet
1472:693993230313091072
Succeeded in retrieving this tweet
1473:693942351086120961
Succeeded in retrieving this tweet
1474:693647888581312512
Succeeded in retrieving this tweet
1475:693644216740769793
Succeeded in retrieving this tweet
1476:6936422321512857

Succeeded in retrieving this tweet
1599:686035780142297088
Succeeded in retrieving this tweet
1600:686034024800862208
Succeeded in retrieving this tweet
1601:686007916130873345
Succeeded in retrieving this tweet
1602:686003207160610816
Succeeded in retrieving this tweet
1603:685973236358713344
Succeeded in retrieving this tweet
1604:685943807276412928
Succeeded in retrieving this tweet
1605:685906723014619143
Succeeded in retrieving this tweet
1606:685681090388975616
Succeeded in retrieving this tweet
1607:685667379192414208
Succeeded in retrieving this tweet
1608:685663452032069632
Succeeded in retrieving this tweet
1609:685641971164143616
Succeeded in retrieving this tweet
1610:685547936038666240
Succeeded in retrieving this tweet
1611:685532292383666176
Succeeded in retrieving this tweet
1612:685325112850124800
Succeeded in retrieving this tweet
1613:685321586178670592
Succeeded in retrieving this tweet
1614:685315239903100929
Succeeded in retrieving this tweet
1615:6853074517013340

Succeeded in retrieving this tweet
1738:679530280114372609
Succeeded in retrieving this tweet
1739:679527802031484928
Succeeded in retrieving this tweet
1740:679511351870550016
Succeeded in retrieving this tweet
1741:679503373272485890
Succeeded in retrieving this tweet
1742:679475951516934144
Succeeded in retrieving this tweet
1743:679462823135686656
Succeeded in retrieving this tweet
1744:679405845277462528
Succeeded in retrieving this tweet
1745:679158373988876288
Succeeded in retrieving this tweet
1746:679148763231985668
Succeeded in retrieving this tweet
1747:679132435750195208
Succeeded in retrieving this tweet
1748:679111216690831360
Succeeded in retrieving this tweet
1749:679062614270468097
Succeeded in retrieving this tweet
1750:679047485189439488
Succeeded in retrieving this tweet
1751:679001094530465792
Succeeded in retrieving this tweet
1752:678991772295516161
Succeeded in retrieving this tweet
1753:678969228704284672
Succeeded in retrieving this tweet
1754:6788002836490690

Rate limit reached. Sleeping for: 290


Succeeded in retrieving this tweet
1802:676957860086095872
Succeeded in retrieving this tweet
1803:676949632774234114
Succeeded in retrieving this tweet
1804:676948236477857792
Succeeded in retrieving this tweet
1805:676946864479084545
Succeeded in retrieving this tweet
1806:676942428000112642
Succeeded in retrieving this tweet
1807:676936541936185344
Succeeded in retrieving this tweet
1808:676916996760600576
Succeeded in retrieving this tweet
1809:676897532954456065
Succeeded in retrieving this tweet
1810:676864501615042560
Succeeded in retrieving this tweet
1811:676821958043033607
Succeeded in retrieving this tweet
1812:676819651066732545
Succeeded in retrieving this tweet
1813:676811746707918848
Succeeded in retrieving this tweet
1814:676776431406465024
Succeeded in retrieving this tweet
1815:676617503762681856
Succeeded in retrieving this tweet
1816:676613908052996102
Succeeded in retrieving this tweet
1817:676606785097199616
Succeeded in retrieving this tweet
1818:6766033933145784

Succeeded in retrieving this tweet
1941:673716320723169284
Succeeded in retrieving this tweet
1942:673715861853720576
Succeeded in retrieving this tweet
1943:673711475735838725
Succeeded in retrieving this tweet
1944:673709992831262724
Succeeded in retrieving this tweet
1945:673708611235921920
Succeeded in retrieving this tweet
1946:673707060090052608
Succeeded in retrieving this tweet
1947:673705679337693185
Succeeded in retrieving this tweet
1948:673700254269775872
Succeeded in retrieving this tweet
1949:673697980713705472
Succeeded in retrieving this tweet
1950:673689733134946305
Succeeded in retrieving this tweet
1951:673688752737402881
Succeeded in retrieving this tweet
1952:673686845050527744
Succeeded in retrieving this tweet
1953:673680198160809984
Succeeded in retrieving this tweet
1954:673662677122719744
Succeeded in retrieving this tweet
1955:673656262056419329
Succeeded in retrieving this tweet
1956:673636718965334016
Succeeded in retrieving this tweet
1957:6736128540801966

Succeeded in retrieving this tweet
2080:670826280409919488
Succeeded in retrieving this tweet
2081:670823764196741120
Succeeded in retrieving this tweet
2082:670822709593571328
Succeeded in retrieving this tweet
2083:670815497391357952
Succeeded in retrieving this tweet
2084:670811965569282048
Succeeded in retrieving this tweet
2085:670807719151067136
Succeeded in retrieving this tweet
2086:670804601705242624
Succeeded in retrieving this tweet
2087:670803562457407488
Succeeded in retrieving this tweet
2088:670797304698376195
Succeeded in retrieving this tweet
2089:670792680469889025
Succeeded in retrieving this tweet
2090:670789397210615808
Succeeded in retrieving this tweet
2091:670786190031921152
Succeeded in retrieving this tweet
2092:670783437142401025
Succeeded in retrieving this tweet
2093:670782429121134593
Succeeded in retrieving this tweet
2094:670780561024270336
Succeeded in retrieving this tweet
2095:670778058496974848
Succeeded in retrieving this tweet
2096:6707641036239667

Succeeded in retrieving this tweet
2219:668507509523615744
Succeeded in retrieving this tweet
2220:668496999348633600
Succeeded in retrieving this tweet
2221:668484198282485761
Succeeded in retrieving this tweet
2222:668480044826800133
Succeeded in retrieving this tweet
2223:668466899341221888
Succeeded in retrieving this tweet
2224:668297328638447616
Succeeded in retrieving this tweet
2225:668291999406125056
Succeeded in retrieving this tweet
2226:668286279830867968
Succeeded in retrieving this tweet
2227:668274247790391296
Succeeded in retrieving this tweet
2228:668268907921326080
Succeeded in retrieving this tweet
2229:668256321989451776
Succeeded in retrieving this tweet
2230:668248472370458624
Succeeded in retrieving this tweet
2231:668237644992782336
Succeeded in retrieving this tweet
2232:668226093875376128
Succeeded in retrieving this tweet
2233:668221241640230912
Succeeded in retrieving this tweet
2234:668204964695683073
Succeeded in retrieving this tweet
2235:6681906814463795

NameError: name 'end' is not defined

In [11]:
twitter_posts.head(50)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0   892420643555336193                    NaN                  NaN   
1   892177421306343426                    NaN                  NaN   
2   891815181378084864                    NaN                  NaN   
3   891689557279858688                    NaN                  NaN   
4   891327558926688256                    NaN                  NaN   
5   891087950875897856                    NaN                  NaN   
6   890971913173991426                    NaN                  NaN   
7   890729181411237888                    NaN                  NaN   
8   890609185150312448                    NaN                  NaN   
9   890240255349198849                    NaN                  NaN   
10  890006608113172480                    NaN                  NaN   
11  889880896479866881                    NaN                  NaN   
12  889665388333682689                    NaN                  NaN   
13  889638837579907072                    NaN                  NaN   
14  889531135344209921                    NaN                  NaN   
15  889278841981685760                    NaN                  NaN   
16  888917238123831296                    NaN                  NaN   
17  888804989199671297                    NaN                  NaN   
18  888554962724278272                    NaN                  NaN   
19  888202515573088257                    NaN                  NaN   
20  888078434458587136                    NaN                  NaN   
21  887705289381826560                    NaN                  NaN   
22  887517139158093824                    NaN                  NaN   
23  887473957103951883                    NaN                  NaN   
24  887343217045368832                    NaN                  NaN   
25  887101392804085760                    NaN                  NaN   
26  886983233522544640                    NaN                  NaN   
27  886736880519319552                    NaN                  NaN   
28  886680336477933568                    NaN                  NaN   
29  886366144734445568                    NaN                  NaN   
30  886267009285017600           8.862664e+17         2.281182e+09   
31  886258384151887873                    NaN                  NaN   
32  886054160059072513                    NaN                  NaN   
33  885984800019947520                    NaN                  NaN   
34  885528943205470208                    NaN                  NaN   
35  885518971528720385                    NaN                  NaN   
36  885311592912609280                    NaN                  NaN   
37  885167619883638784                    NaN                  NaN   
38  884925521741709313                    NaN                  NaN   
39  884876753390489601                    NaN                  NaN   
40  884562892145688576                    NaN                  NaN   
41  884441805382717440                    NaN                  NaN   
42  884247878851493888                    NaN                  NaN   
43  884162670584377345                    NaN                  NaN   
44  883838122936631299                    NaN                  NaN   
45  883482846933004288                    NaN                  NaN   
46  883360690899218434                    NaN                  NaN   
47  883117836046086144                    NaN                  NaN   
48  882992080364220416                    NaN                  NaN   
49  882762694511734784                    NaN                  NaN   

                    timestamp  \
0   2017-08-01 16:23:56 +0000   
1   2017-08-01 00:17:27 +0000   
2   2017-07-31 00:18:03 +0000   
3   2017-07-30 15:58:51 +0000   
4   2017-07-29 16:00:24 +0000   
5   2017-07-29 00:08:17 +0000   
6   2017-07-28 16:27:12 +0000   
7   2017-07-28 00:22:40 +0000   
8   2017-07-27 16:25:51 +0000   
9   2017-07-26 15:59:51 +0000   
10  2017-07-26 00:31:25 +0000   
11  2017-07-25 16:11:53 +0000   
12  2017-07-25

In [12]:
twitter_posts.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [13]:
twitter_posts.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [14]:
img_predict.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [15]:
img_predict.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

To check the percentage of null values

In [16]:
img_predict.isnull().sum()/img_predict.shape[0]*100

tweet_id    0.0
jpg_url     0.0
img_num     0.0
p1          0.0
p1_conf     0.0
p1_dog      0.0
p2          0.0
p2_conf     0.0
p2_dog      0.0
p3          0.0
p3_conf     0.0
p3_dog      0.0
dtype: float64

In [17]:
twitter_posts.isnull().sum()/twitter_posts.shape[0]*100

tweet_id                       0.000000
in_reply_to_status_id         96.689304
in_reply_to_user_id           96.689304
timestamp                      0.000000
source                         0.000000
text                           0.000000
retweeted_status_id           92.317487
retweeted_status_user_id      92.317487
retweeted_status_timestamp    92.317487
expanded_urls                  2.504244
rating_numerator               0.000000
rating_denominator             0.000000
name                           0.000000
doggo                          0.000000
floofer                        0.000000
pupper                         0.000000
puppo                          0.000000
dtype: float64

### Quality issues

1. Rename the 'name' column to 'Dog names'.

2. The timestamp column data type is supposed to be datetime not object.

3. The in_reply_to_status_id column contains about 96.689 percent of null values.

4. The in_reply_to_user_id column also contains about 96.689 percent of null values ie NaN.

5. The retweeted_status_id has about 92.3 percent of null values.

6. The retweeted_status_user_id also contains about 92.317 percent of null values.

7. Columns doggo, floofer, pupper, puppo have to be merged into one column under the name column.

8. None observation/rows has to be replaced with nan.




### Tidiness issues
1. Under the image data, column names p1, p2, p3 are not descriptive enough, it has to be changed.

2. The retweeted_status_timestamp column is not an object but a data/time module.

## Cleaning Data

Before cleaning, i have tio make a deep copy of my datasets.

In [24]:
# Make copies of original pieces of data
twitter_posts_copy = twitter_posts.copy(deep=True)
img_predict_copy = img_predict.copy(deep=True)

### Issue #1: The column name 'name' is not decriptive enough

# Define:  Rename the 'name' column to 'Dog names'

# Code

1. To see the columns in our twitter image dataset

2. Make a deep copy of our datasets.

3. Change column name from 'name' to 'Dog names'.

In [25]:
img_predict_copy.columns

Index(['tweet_id', 'jpg_url', 'img_num', 'p1', 'p1_conf', 'p1_dog', 'p2',
       'p2_conf', 'p2_dog', 'p3', 'p3_conf', 'p3_dog'],
      dtype='object')

In [27]:
twitter_posts_copy.columns

Index(['tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp',
       'source', 'text', 'retweeted_status_id', 'retweeted_status_user_id',
       'retweeted_status_timestamp', 'expanded_urls', 'rating_numerator',
       'rating_denominator', 'name', 'doggo', 'floofer', 'pupper', 'puppo'],
      dtype='object')

# Test

In [28]:
twitter_posts_copy.rename(columns = {'name': 'Dog names'})

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
...                  ...                    ...                  ...   
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
0     2017-08-01 16:23:56 +0000   
1     2017-08-01 00:17:27 +0000   
2     2017-07-31 00:18:03 +0000   
3     2017-07-30 15:58:51 +0000   
4     2017-07-29 16:00:24 +0000   
...                         ...   
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
0     This is Phineas. He's a mystical boy. Only eve...                  NaN   
1     This is Tilly. She's just checking pup on you....                  NaN   
2     This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3     This is Darla. She commenced a snooze mid meal...                  NaN   
4     This is Franklin. He would like you to stop ca...                  NaN   
...                                                 ...                  ...   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
0                          NaN                        NaN   
1                          NaN                        NaN   
2                          NaN                        NaN   
3                          NaN                        NaN   
4                          NaN                        NaN   
...                        ...                        ...   
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
0     https://twitter.com/dog_rates/status/892420643...                13   
1     https://twitter.com/dog_rates/status/892177421...                13   
2     https://twitter.com/dog_rates/status/891815181...                12   
3     https://twitt

The column name has been changed from name to Dog names. It has been cleaned.

## Issue 2

Define: Split the timestamp column into three different columns by delimiter.

Code

In [85]:
twitter_posts_copy[['Date', 'Time', 'Stamp']] = twitter_posts_copy['timestamp'].str.split(' ', expand=True)

Test

In [86]:
twitter_posts_copy.columns

Index(['tweet_id', 'timestamp', 'source', 'text', 'retweeted_status_id',
       'retweeted_status_user_id', 'retweeted_status_timestamp',
       'expanded_urls', 'rating_numerator', 'rating_denominator', 'name',
       'doggo', 'floofer', 'pupper', 'puppo', 'Date', 'Time', 'Stamp'],
      dtype='object')

# Issue 3: Change the datatype of the timestamp column from object to datetime

# Define

- Convert the datatype of timestamp from 'object' into 'datetime'.


# Code


In [90]:
twitter_posts_copy.head(2)

tweet_id                  timestamp  \
0  892420643555336193  2017-08-01 16:23:56 +0000   
1  892177421306343426  2017-08-01 00:17:27 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   

   rating_denominator     name doggo floofer pupper puppo        Date  \
0                  10  Phineas  None    None   None  None  2017-08-01   
1                  10    Tilly  None    None   None  None  2017-08-01   

       Time  Stamp  
0  16:23:56  +0000  
1  00:17:27  +0000

In [91]:
twitter_posts_copy['Date'] = twitter_posts_copy['Date'].apply(lambda x: datetime.datetime.strptime(x, '%Y-%m-%d'))

Test

In [92]:
twitter_posts_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 18 columns):
tweet_id                      2356 non-null int64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
Date                          2356 non-null datetime64[ns]
Time                          2356 non-null object
Stamp                 

Issue 4

Define

To drop some unuseful columns like (Date, Time, Stamp)

Code

In [59]:
twitter_posts_copy.drop(('Date', 'Time', 'Stamp'), axis=1, inplace=True)

Test

In [61]:
twitter_posts_copy.head(1)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   

   rating_denominator     name doggo floofer pupper puppo       Date  \
0                  10  Phineas  None    None   None  None 2017-08-01   

       Time  Stamp  
0  16:23:56  +0000

# Code

In [42]:
twitter_posts_copy[['Date', 'Time', 'Stamp']] = twitter_posts_copy['timestamp'].str.split(' ', expand=True)

In [41]:
twitter_posts_copy.head(2)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   

                                       expanded_urls  ...  rating_denominator  \
0  https://twitter.com/dog_rates/status/892420643...  ...                  10   
1  https://twitter.com/dog_rates/status/892177421...  ...                  10   

      name doggo floofer pupper puppo            (Date, Time, Stamp)  \
0  Phineas  None    None   None  None  [2017-08-01, 16:23:56, +0000]   
1    Tilly  None    None   None  None  [2017-08-01, 00:17:27, +0000]   

         Date      Time  Stamp  
0  2017-08-01  16:23:56  +0000  
1  2017-08-01  00:17:27  +0000  

[2 rows x 21 columns]

Issue 4

Define
- Drop the in_reply_to_status_id column because it has over 70% null values.

Code

In [70]:
twitter_posts_copy.drop(('in_reply_to_status_id'), axis=1, inplace=True)

Test

In [71]:
twitter_posts_copy.head(2)

tweet_id  in_reply_to_user_id                  timestamp  \
0  892420643555336193                  NaN  2017-08-01 16:23:56 +0000   
1  892177421306343426                  NaN  2017-08-01 00:17:27 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   

   rating_denominator     name doggo floofer pupper puppo       Date  \
0                  10  Phineas  None    None   None  None 2017-08-01   
1                  10    Tilly  None    None   None  None 2017-08-01   

       Time  Stamp  
0  16:23:56  +0000  
1  00:17:27  +0000

Issue 5

Define
- Drop/delete the in_reply_to_user_id column due to over 90% of its rows are null.

Code

In [72]:
twitter_posts_copy.drop(('in_reply_to_user_id'), axis=1, inplace=True)

Test

In [73]:
twitter_posts_copy

tweet_id                  timestamp  \
0     892420643555336193  2017-08-01 16:23:56 +0000   
1     892177421306343426  2017-08-01 00:17:27 +0000   
2     891815181378084864  2017-07-31 00:18:03 +0000   
3     891689557279858688  2017-07-30 15:58:51 +0000   
4     891327558926688256  2017-07-29 16:00:24 +0000   
...                  ...                        ...   
2351  666049248165822465  2015-11-16 00:24:50 +0000   
2352  666044226329800704  2015-11-16 00:04:52 +0000   
2353  666033412701032449  2015-11-15 23:21:54 +0000   
2354  666029285002620928  2015-11-15 23:05:30 +0000   
2355  666020888022790149  2015-11-15 22:32:08 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
0     This is Phineas. He's a mystical boy. Only eve...                  NaN   
1     This is Tilly. She's just checking pup on you....                  NaN   
2     This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3     This is Darla. She commenced a snooze mid meal...                  NaN   
4     This is Franklin. He would like you to stop ca...                  NaN   
...                                                 ...                  ...   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
0                          NaN                        NaN   
1                          NaN                        NaN   
2                          NaN                        NaN   
3                          NaN                        NaN   
4                          NaN                        NaN   
...                        ...                        ...   
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
0     https://twitter.com/dog_rates/status/892420643...                13   
1     https://twitter.com/dog_rates/status/892177421...                13   
2     https://twitter.com/dog_rates/status/891815181...                12   
3     https://twitter.com/dog_rates/status/891689557...                13   
4     https://twitter.com/dog_rates/status/891327558...                12   
...                                                 ...               ...   
2351  https://twitter.com/dog_rates/status/666049248...                 5   
2352  https://twitter.com/dog_rates/status/666044226...                 6   
2353  https://twitter.com/dog_rates/status/666033412...                 9   
2354  https://twitter.com/dog_rates/status/666029285...                 7   
2355  https://twitter.com/dog_rates/status/666020888...                 8   

      rating_denominator   